lib/services/role.go

```diff
@@ -1075,25 +1087,35 @@ func (set RoleSet) CheckKubeGroupsAndUsers(ttl time.Duration, overrideTTL bool,
 	for _, role := range set {
 		ok, _, err := RoleMatchers(matchers).MatchAny(role, types.Deny)
 		if err != nil {
-			return nil, nil, trace.Wrap(err)
+			return KubePermissions{}, trace.Wrap(err)
 		}
 		if !ok {
 			continue
 		}
 		for _, group := range role.GetKubeGroups(types.Deny) {
+			if group == teleport.KubeSystemAuthenticated {
+				denyAuthenticated = true
+			}
+
 			delete(groups, group)
 		}
 		for _, user := range role.GetKubeUsers(types.Deny) {
 			delete(users, user)
 		}
 	}
-	if !matchedTTL {
-		return nil, nil, trace.AccessDenied("this user cannot request kubernetes access for %v", ttl)
+
+	if !matchedTTL && !denyAuthenticated {
+		return KubePermissions{}, trace.AccessDenied("this user cannot request kubernetes access for %v", ttl)
 	}
-	if len(groups) == 0 && len(users) == 0 {
-		return nil, nil, trace.NotFound("this user cannot request kubernetes access, has no assigned groups or users")
+	if len(groups) == 0 && len(users) == 0 && !denyAuthenticated {
```

Comment:
Why do these checks require `&& !denyAuthenticated` now?

Reply:
Because otherwise you always need to assign some role or user which may not be what you want when running with GKE Groups RBAC plugin. Without this, a hypothetical scenario when you rely only on that plugin will force you to assign a dummy user (adding a group would prevent it from fetching the Google Groups RBAC stuff) to get past this check.

Reply:
These checks exist to prevent people with any roles at all from using the system:authenticated role to query information.

Reply:
But if you explicitly add a deny rule for that you want the check to go through anyway since we no longer append the system:authenticated group anyway (so you can't use those permissions.

Reply:
I think that rejecting access when you don't have either kube_users or kube_groups assigned is the correct behavior. Otherwise we're making too many assumptions about the scenario which may result in some unforeseen security issue. I would keep the existing behavior and only keep the part of the change that filters out the system:auth group and keep these checks the same. The customer who needs this feature does specify kube_users AFAIK.

Reply:
This doesn't feel like it makes total sense though. While we could reject here, this still doesn't grant additional access. The only reason we have these checks is to guard against people without any kube roles having `system:authenticated` permissions. I don't think we're making any assumptions here that are noteworthy since it doesn't change anything access-wise.

Reply:
That's my view but if you still disagree I'd be a happy to change it to early reject within Teleport.

Reply:
Yes, to be on the safe side I would not expand the scope of this (very specific case) change and keep these checks as they are. We can revisit if it still causes issues for someone.